ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Mempool/Impl/Common.hs

```diff
@@ -279,17 +279,25 @@ validateNewTransaction
      , InternalState blk
      )
 validateNewTransaction cfg wti tx txsz st is =
-    case runExcept (applyTx cfg wti isSlotNo tx st) of
+    case runExcept res of
       Left err         -> ( Left err, is )
       Right (st', vtx) ->
         ( Right vtx
         , is { isTxs          = isTxs :> TxTicket vtx nextTicketNo txsz
-             , isTxIds        = Set.insert (txId tx) isTxIds
+             , isTxIds        = Map.insert (txId tx) vtx isTxIds
              , isLedgerState  = prependDiffs isLedgerState st'
              , isLastTicketNo = nextTicketNo
              }
         )
   where
+    res = case Map.lookup (txId tx) isTxIds of
+      Nothing  -> applyTx cfg wti isSlotNo tx st
+      Just vtx -> do
+        _ <- reapplyTx cfg isSlotNo vtx st
+        -- that must have failed, since every tx must consume at least one
+        -- input and this one is necessarily a double-spend
+        pure $ error "unreachable"
```

Comment:
Am I reading this wrong, or will this only fire when double-adding transactions into the mempool? That will be indeed a fail-fast situation, but it won't help when the tx is not in the mempool but instead has already been applied to the state, no?

Reply:
> Am I reading this wrong, or will this only fire when double-adding transactions into the mempool?

Yes, this will indeed only fire in the scenario you describe.

> but it won't help when the tx is not in the mempool but instead has already been applied to the state, no?

Exactly :+1: For the purpose of this PR (benchmarking when/why tx application is failing slowly) this is enough and was convenient and very simple to implement, but we indeed want the final optimization to be more general, see https://github.com/IntersectMBO/cardano-ledger/issues/4849#issuecomment-2607834134.